jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHttpOutputInterceptor.java

```diff
@@ -325,82 +311,115 @@ protected void onCompleteFailure(Throwable x)
+                BufferUtil.fill(_buffer, GZIP_HEADER, 0, GZIP_HEADER.length);
+            }
+            else
             {
-                if (BufferUtil.isEmpty(_content))
+                // otherwise clear the buffer as previous writes will always fully consume.
+                BufferUtil.clear(_buffer);
+            }
+
+            // If the deflator is not finished, then compress more data
+            if (!_deflater.finished())
+            {
+                if (_deflater.needsInput())
                 {
-                    if (_deflater.finished())
+                    // if there is no more content available to compress
+                    // then we are either finished all content or just the current write.
+                    if (BufferUtil.isEmpty(_content))
                     {
-                        _factory.recycle(_deflater);
-                        _deflater = null;
-                        _channel.getByteBufferPool().release(_buffer);
-                        _buffer = null;
-                        if (_copy != null)
+                        if (_last)
+                            _deflater.finish();
+                        else
+                            return Action.SUCCEEDED;
+                    }
+                    else
+                    {
+                        // If there is more content available to compress, we have to make sure
+                        // it is available in an array for the current deflator API, maybe slicing
+                        // of content.
+                        ByteBuffer slice;
+                        if (_content.hasArray())
+                            slice = _content;
+                        else
                         {
-                            _channel.getByteBufferPool().release(_copy);
-                            _copy = null;
+                            if (_copy == null)
+                                _copy = _channel.getByteBufferPool().acquire(_bufferSize, false);
+                            else
+                                BufferUtil.clear(_copy);
+                            slice = _copy;
+                            int copied = BufferUtil.append(_copy, _content);
                         }
-                        return Action.SUCCEEDED;
-                    }
 
-                    if (!_last)
-                    {
-                        return Action.SUCCEEDED;
+                        // transfer the data from the slice to the the deflator
+                        byte[] array = slice.array();
+                        int off = slice.arrayOffset() + slice.position();
+                        int len = slice.remaining();
+                        _crc.update(array, off, len);
+                        _deflater.setInput(array, off, len);
+                        BufferUtil.clear(slice);
+                        if (_last && BufferUtil.isEmpty(_content))
```

Comment:
Could `BufferUtil.isEmpty(_content)` have changed since the check at line 353?

Reply:
Yes - the slice could be == to _content, so the line above can empty it, otherwise the branch above slices bytes off _content after 353

Reply:
Is this necessary then, looks like it is always fully consuming `_content`, either the `BufferUtil.clear(slice)` will clear it, or `BufferUtil.append(_copy, _content)` will fully consume it.

Reply:
It is necessary.  The `BufferUtil.append(_copy, _content)` will not fully consume if the copy buffer is smaller than the content.

Reply:
They are always the same size, they are both created to `_bufferSize` which is final.

Reply:
@lachlan-roberts the `_content` buffer is passed in and can be any size.  Specifically it can be of unlimited huge size if a file mapped buffer is used.  This is tested with `org.eclipse.jetty.server.handler.gzip.GzipContentLengthTest#testHttpOutputWrite`

Reply:
Ahh I see, I was confused between `_buffer` and `_content`, you're right.